File: src/converter/plugins/convertNodeWithTypeArguments.ts
import {NodeWithTypeArguments} from "typescript";
import {ifPresent} from "../render.js";
import {createSimplePlugin} from "../plugin.js";

export const convertNodeWithTypeArguments = createSimplePlugin<NodeWithTypeArguments>((node, context, render) =>  {
    const typeArguments = node.typeArguments
        ?.map(typeArgument => render(typeArgument))
        ?.filter(Boolean)
        ?.join(", ")

    return ifPresent(typeArguments, it => `<${it}>`)
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertNodeWithTypeArguments.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.extension.ifPresent
import typescript.Node
import typescript.NodeWithTypeArguments
import typescript.asArray

fun convertNodeWithTypeArguments(node: NodeWithTypeArguments, render: Render<Node>): String {
    val typeArguments = node.typeArguments?.asArray()
        ?.map { render(it) }
        ?.filter { it.isNotEmpty() }
        ?.joinToString(separator = ", ")

    return ifPresent(typeArguments) { "<${it}>" }
}
